transport/src/main/java/io/scalecube/transport/Transport.java

```diff
@@ -323,29 +350,33 @@ public void onMessage(Message message) {
     incomingMessagesSubject.onNext(message);
   }
 
-  private TransportChannel getOrConnect(@CheckForNull InetSocketAddress address) {
+  private ListenableFuture<TransportChannel> getOrConnect(@CheckForNull InetSocketAddress address) {
     checkArgument(address != null);
-    return connectedChannels.get(address, new Computable<InetSocketAddress, TransportChannel>() {
+    return Futures.transform(resolveSocketAddress(address), new Function<InetSocketAddress, TransportChannel>() {
       @Override
-      public TransportChannel compute(final InetSocketAddress address) {
-        final Channel channel = createConnectorChannel();
-        final TransportChannel transportChannel = createConnectorTransportChannel(channel, address);
-
-        LOGGER.info("Registered connector: {}", transportChannel);
-
-        final ChannelFuture registerChannelFuture = eventLoop.register(channel);
-        registerChannelFuture.addListener(new ChannelFutureListener() {
+      public TransportChannel apply(final InetSocketAddress resolvedAddress) {
+        return connectedChannels.get(resolvedAddress, new Computable<InetSocketAddress, TransportChannel>() {
           @Override
-          public void operationComplete(ChannelFuture future) throws Exception {
-            if (future.isSuccess()) {
-              connect(channel, address, transportChannel);
-            } else {
-              channel.unsafe().closeForcibly();
-              transportChannel.close();
-            }
+          public TransportChannel compute(final InetSocketAddress input) {
+            final Channel channel = createConnectorChannel();
+            final TransportChannel transportChannel = createConnectorTransportChannel(channel, input);
+
+            Futures.addCallback(registerChannel(channel), new FutureCallback<Void>() {
```

Comment:
Lets return back previous version of this code since it was more compact; and basically exactly the same.